packages/headless/src/controllers/commerce/context/cart/headless-cart.ts

```diff
@@ -147,13 +148,55 @@ export function buildCart(engine: CommerceEngine, props: CartProps = {}): Cart {
     dispatch(setItems(initialState.items));
   }
 
-  function isNewQuantityDifferent(productId: string, quantity: number) {
-    const item = itemSelector(getState(), productId);
-    if (!item) {
-      return quantity > 0;
-    }
+  function isNewQuantityDifferent(
+    currentItem: CartItem,
+    prevItem: CartItemWithMetadata
+  ) {
+    return prevItem ? prevItem.quantity !== currentItem.quantity : true;
+  }
+
+  function isCurrentItemQuantityGreater(
+    currentItem: CartItem,
+    prevItem: CartItemWithMetadata | undefined
+  ): boolean {
+    return !prevItem || currentItem.quantity > prevItem.quantity;
+  }
+
+  // @todo LENS-1589: currently, the currency attribute is a string. However, the type should be CurrencyCodeISO4217
+  function getCurrency(): CurrencyCodeISO4217 {
+    return engine.state.commerceContext.currency as CurrencyCodeISO4217;
+  }
 
-    return item.quantity !== quantity;
+  function isEqual(
+    currentItem: CartItem,
+    prevItem: CartItemWithMetadata | undefined
+  ): boolean {
+    return prevItem
+      ? currentItem.name === prevItem.name &&
+          currentItem.price === prevItem.price &&
+          currentItem.quantity === prevItem.quantity
+      : false;
+  }
+
+  function createEcCartActionPayload(
+    currentItem: CartItem,
+    prevItem: CartItemWithMetadata | undefined
+  ): Ec.CartAction {
+    const {quantity: currentQuantity, ...product} = currentItem;
+    const action = isCurrentItemQuantityGreater(currentItem, prevItem)
```

Comment:
if `isCurrentItemQuantityGreater` is solely used there, I would make it return add or remove and rename it accordingly, this would make the code more readable.